### PR TITLE
Add a special `Mod` modifier that translates to either `Meta` on Mac or `Control` on other platforms

### LINKED ISSUE #56: Add a special `Mod` character that translates to either `Meta` on Mac or `Control` on other platforms
There are many cases where we want to be able to support _either_ the `Meta` or `Control`  modifier for a shortcut; as opposed to cases where we support _both_ `Meta` and `Control`.

One way to handle this would be to add a special `Mod` character that would map to `Meta` on Mac / iOS, and `Control` on Window or Linux.

```html
<button data-hotkey="Mod+s">Save</button>
```

would map to `Meta+s` on Mac / `Control+s` otherwise.

### PR DESCRIPTION
**Pull request by theinterned**
resolves #56

## What is this?

On Mac, `Meta` (command) is frequently used as the modifier for shortcuts: on Windows and Linux `Control` is used instead. Therefor in order to correctly localize hotkeys, it's common to see code that handles swapping platform-specific modifier keys in users of this library. It would make sense to do this in the library and eliminate all that custom code!

### `Mod` modifier

This PR adds support for using a special `Mod` modifier that automatically localizes to `Control` on Windows / Linux, or `Meta` on `Mac`.

The rational is that this allows an easy way to define cross-platform shortcuts. Without this feature, an implementer would have to either support both `Control` and `Meta` forms of shortcuts, or write their own code to handle swapping these.

### Consistent sorting of modifiers

Although it was only documented in #62 , hotkey has always had a hard requirement that modifiers are sorted in the order that matches the order `eventToHotkeyString` formats the modifier keys from an event.

Since the `Control` and `Meta` keys appear in different places in the hotkey string, in order to support the `Mod` key, I also had to ensure that modifiers in a hotkey string are consistently sorted before comparoson to an event that is serialized via `eventToHotkeyString`.

So this PR also removes the hard requirement that modifiers be sorted in a strict order.

## Approach

I chose to add a new `normalizeHotkey` function to handle localizing the `Mod` key _and_ sorting the modifiers. This `normalizeHotkey` is applied to the hotkey string in the `expandHotkeyToEdges` before it is added to the Trie. This ensures that when we later look for a serialized event in the trie, the hotkey string is already normalized to match the serialized output of `eventToHotkeyString`.

## Limitations

The `"Mod"` key is not currently compatible with `Shift` and `Alt` modifiers.

Because Mac treats `event.key` differently than Windows or Linux when the `Shift` and / or `Alt` key is pressed, it means that the `Mod` modifier will not localize as expected when paired with `Shift`. I have noted this limitation in the [tests](https://github.com/github/hotkey/pull/66#discussion_r777061585) and [in the README](https://github.com/github/hotkey/pull/66#discussion_r777062118). I was hoping to address #54 and #68 in this PR too but decided to defer to keep the size of the PR down.

## TODO

- [x] implement `normalizeHotkey` function to allow comparing a `eventToHotkey` string to a hotkey string containing a `Mod` character.
    - [x] Account for hotkeys that contain both `Control` and `Meta`!
    - [ ] Account for how Mac lowercases `event.key` when `Meta+Shift` are pressed.
- [x] Make `normalizeHotkey ` platform specific: convert `Mod` to `Meta` on Mac  or `Control` on other platforms.
- [x] Use this `normalizeHotkey ` to map hotkey to Trie in `expandHotkeyToEdges` so that the Trie contains the platform-specific modifier: use it to allow `keyDownHandler` to match events to hotkey strings containing the `Mod` alias.
- [x] Add docs about `Mod` character
    - [x] Add special notes about `Shift` etc?

### CODE REVIEW

File: examples/hotkey_mapper.html [hunk 1]
```diff
@@ -1,45 +1,61 @@
-<!doctype html>
```
Comment (theinterned) [inline examples/hotkey_mapper.html:1]: there's no functional change — this is just the result form running prettier on this file.

File: examples/index.html [hunk 2]
```diff
@@ -1,29 +1,33 @@
-<!doctype html>
```
Comment (theinterned) [inline examples/index.html:1]: same — there's no functional change — this is just the result form running prettier on this file.

File: src/utils.ts [hunk 3]
```diff
@@ -25,5 +27,5 @@ export function fireDeterminedAction(el: HTMLElement, path: string[]): void {
 }
 
 export function expandHotkeyToEdges(hotkey: string): string[][] {
-  return hotkey.split(',').map(edge => edge.split(' '))
+  return hotkey.split(',').map(edge => edge.split(' ').map(k => normalizeHotkey(k)))
```
Comment (theinterned) [inline src/utils.ts:30]: this is where the normalization is applied, so we take the user-provided hotkey string, normalize (or localize?) it for the current platform and store that on the trie. Then when we later serialize a keyboard event we compare it to the normalized version of the string.

File: test/test-normalize-hotkey.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    ['Control+Meta+a', 'Control+Meta+a'],
+    // Mod should be localized based on platform
+    ['Mod+a', 'Control+a', 'win'],
+    ['Mod+a', 'Meta+a', 'mac'],
+    ['Mod+A', 'Control+A', 'win'],
+    ['Mod+A', 'Meta+A', 'mac'], // TODO: on a mac upper-case keys are lowercased when Meta is pressed
+    ['Mod+9', 'Control+9', 'win'],
+    ['Mod+9', 'Meta+9', 'mac'],
+    ['Mod+)', 'Control+)', 'win'],
+    ['Mod+)', 'Meta+)', 'mac'], // TODO: on a mac upper-case keys are lowercased when Meta is pressed
```
Comment (theinterned) [inline test/test-normalize-hotkey.js:22]: Note that the tests on line 18 and 22 are incorrect: the hotkey `Meta+A` won't match on Mac: instead it _should_ be `Meta+Shift+a` in the current  implementation. Likewise, `Meta+)` should be `Meta+Shift+0`.

I hope to address this in https://github.com/github/hotkey/issues/54

File: README.md [hunk 5]
```diff
@@ -69,14 +76,19 @@ for (const el of document.querySelectorAll('[data-hotkey]')) {
 2. At minimum a hotkey string must specify one bare key.
 3. Multiple hotkeys (aliases) are separated by a `,`. For example the hotkey `a,b` would activate if the user typed `a` or `b`.
 4. Multiple keys separated by a blank space represent a key sequence. For example the hotkey `g n` would activate when a user types the `g` key followed by the `n` key.
-5. Modifier key combos are separated with a `+` and are prepended to a key in a consistent order as follows: `Control+Alt+Meta+Shift+KEY`.
+5. Modifier key combos are separated with a `+` and are prepended to a key in a consistent order as follows: `"Control+Alt+Meta+Shift+KEY"`.
+6. `"Mod"` is a special modifier that localizes to `Meta` on MacOS/iOS, and `Control` on Windows/Linux.
+   1. `"Mod+"` can appear in any order in a hotkey string. For example: `"Mod+Alt+Shift+KEY"`
+   2. Neither the `Control` or `Meta` modifiers should appear in a hotkey string with `Mod`.
+   3. Due to the inconsistent lower-caseing of `event.key` on Mac and iOS when `Meta` is pressed along with `Shift`, it is recommended to avoid hotkey strings containing both `Mod` and `Shift`.
```
Comment (theinterned) [inline README.md:83]: I've added this note for now, but hope to resolve this inconsistency in #54 — I will make a note to update the README in that issue.

File: test/test-normalize-hotkey.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    ['Mod+Alt+a', 'Control+Alt+a', 'win'],
+    ['Mod+Alt+a', 'Alt+Meta+a', 'mac'],
+    // Modifier sorting
+    ['Shift+Alt+Meta+Control+m', 'Control+Alt+Meta+Shift+m'],
```
Comment (khiga8) [inline test/test-normalize-hotkey.js:26]: I know this test is only meant to demonstrate sorting works but wondering if we should exclude this given it's invalid. Or maybe we can add a note that this is invalid [along with note to not use `Mod` with `Meta` or `Control`](https://github.com/github/hotkey/pull/66/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R82)

**Review by khiga8 — APPROVED** [review]
Thanks for adding the notes around inconsistent behavior 🙇‍♀️

File: test/test-normalize-hotkey.js (hunk 6 above)
Comment (theinterned) [reply test/test-normalize-hotkey.js:26]: I think it is valid to have shortcuts that pair `Control` and `Meta` (just not either of those keys with `Mod`).

Comment (khiga8) [reply test/test-normalize-hotkey.js:26]: You're right, it's technically valid on mac!

File: test/test-normalize-hotkey.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    ['Shift+Alt+Mod+m', 'Control+Alt+Shift+m', 'win']
```
[the remark this answers is not in the document]
Comment (theinterned) [reply test/test-normalize-hotkey.js:27]: I can make this change for sure!

[the remark this answers is not in the document]
Comment (theinterned) [reply test/test-normalize-hotkey.js:30]: Thank you!

I'm not sure if this is what you had in mind for consideration of linux? https://github.com/github/hotkey/pull/66/commits/2a61fcba211a6367a8442cb4d8ff5e305761ff15

I also added a test for the other Mac platform strings https://github.com/github/hotkey/pull/66/commits/9171795ed5a749ae150c45e9fd3ef445cbea3de0

File: pages/demo.html [hunk 8]
```diff
@@ -17,7 +17,8 @@
   <a href="#ok" data-hotkey="o k">Press <kbd>o k</kbd> click this link</a>
 
   <script type="module">
-    import {install} from '../dist/index.js'
+    // import {install} from '../dist/index.js'
+    import {install} from 'https://unpkg.com/@github/hotkey@latest?module'
```
Comment (theinterned) [inline pages/demo.html:21]: I think given the change @iansan5653 just made in https://github.com/github/hotkey/pull/91 this change should no longer apply

File: README.md [hunk 9]
```diff
@@ -97,15 +104,20 @@ for (const el of document.querySelectorAll('[data-shortcut]')) {
...
+7. You can use the comma key `,` as a hotkey, e.g. `a,,` would activate if the user typed `a` or `,`. `Control+,,x` would activate for `Control+,` or `x`.
 
 ### Example
 
 The following hotkey would match if the user typed the key sequence `a` and then `b`, OR if the user held down the `Control`, `Alt` and `/` keys at the same time.
 
 ```js
-"a b,Control+Alt+/"
+'a b,Control+Alt+/'
+
```
Comment (theinterned) [inline README.md:120]:
```suggestion
```

**Review by theinterned — COMMENTED** [review]
I can't approve this because it's my own PR, but there are a few minor tweaks regardless. Besides that I think this all makes sense. Exciting to see this finally getting finalized 😃

File: pages/demo.html (hunk 8 above)
Comment (theinterned) [reply pages/demo.html]: removed this change 6693f1f

**Review by JoyceZhu — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -97,15 +104,19 @@ for (const el of document.querySelectorAll('[data-shortcut]')) {
 2. At minimum a hotkey string must specify one bare key.
 3. Multiple hotkeys (aliases) are separated by a `,`. For example the hotkey `a,b` would activate if the user typed `a` or `b`.
 4. Multiple keys separated by a blank space represent a key sequence. For example the hotkey `g n` would activate when a user types the `g` key followed by the `n` key.
-5. Modifier key combos are separated with a `+` and are prepended to a key in a consistent order as follows: `Control+Alt+Meta+Shift+KEY`.
-6. You can use the comma key `,` as a hotkey, e.g. `a,,` would activate if the user typed `a` or `,`. `Control+,,x` would activate for `Control+,` or `x`.
+5. Modifier key combos are separated with a `+` and are prepended to a key in a consistent order as follows: `"Control+Alt+Meta+Shift+KEY"`.
+6. `"Mod"` is a special modifier that localizes to `Meta` on MacOS/iOS, and `Control` on Windows/Linux.
```
[the remark this answers is not in the document]
Comment (iansan5653) [reply README.md:108]: I think technically Android is Linux, so that might cover it?

[the remark this answers is not in the document]
Comment (JoyceZhu) [reply README.md:108]: oh, right -- that's fair! I've read enough "Android isn't REAL Linux" stuff on the internet I completely forgot about this 😆

**Pull request merged**